rfcs/rfc-story-based-git-workflow.md

```diff
@@ -0,0 +1,117 @@
+
+1. Dev1 and Dev2 have a conversation (probably including the PO or BA), reviewing the goal, the UX design, and technical approach.
+2. Dev1 and Dev2 create a story branch off master (choosing a name that identifies their story). 
+3. Working off a shared design and technical understanding, Dev1 begins the API and data model changes while Dev2 starts working up the screens. They can choose how they wish to commit to their shared branch (they can use pull requests and get each other to do code reviews if they want, do informal "over the shoulder" code reviews , or pair program).
+4. Dev1 and Dev2 work together to update the automated test coverage and ensure the tests all run in their local environment. They review their own code against common, published project coding standards and run any necessary static analysis tests, making corrections as necessary.
+5. As needed, Dev1 and Dev2 compare their branch with Master and rebase as necessary to keep from getting too far out of date. Rebasing early and often is easier than waiting, but it depends on the circumstaces. Dev1 and Dev2 stay in communication daily with the rest of the committers to stay abreast of changes that are coming soon to master.
+6. As needed, Dev1 and Dev2 compare their branch with other story branches that are currently in progress (same reason as above). 
+7. When Dev1 and Dev2 are satisfied that their story is complete and is up-to-date with MASTER, they create a pull request.
+8. The code is reviewed and a few small issues are corrected.
+9. The code is merged into master (squashing commits from the story branch) and the full test suite is re-run in the E2E environment. Because the tests have been run already, there is little chance of breaking the build at this point.
+10. The story branch is cleaned up (destroyed).
+
+# Motivation
+
+The original entities team ("A" team, "Coops" team, team "Aardvark") has been using an approach where they all fork the master branch individually and commit small changes as often as possible. The advantage to this approach is that it keeps merges incremental and small. 
+
+Adding team Bumblebee (FreshWorks) as codebase contributors to the <https://github.com/bcgov/lear> repository exposed some workflow issues around being able to continuously develop with non co-located teams comprised of many team members. The review and commit strategies that were effective with one team do not scale to two teams. There are some other advantages to story branching that can be realized as well.
+
+Here are some of the issues we have seen:
+
+* Frequent small commits that invalidate the test cycle for a story that is almost ready to be created as a pull request
+* Lack of transparency between developers' work in progress (individual developer forks live on GitHub accounts on an ad-hoc basis)
+* Commits to master that are incomplete or break the build
+* Commits to master that do not pass linting
+* Developers committing their own code without a review
+
```

Comment:
Rarely is the build broken ...

Reply:
It would be better to have a process to prevent the build from being broken ever. Checking in a back-end change without a corresponding front end change (or vice versa) will break the build, and this seems to be the current practice unless I am mistaken.

Reply:
No that is incorrect. If you are a monolith, then yes you are moving all portions at one time. But in the case of APIs and possibly many front-ends, this is simply not the case.

Reply:
We are building a single application at this time. There are no components that are being re-used by multiple clients. Release management is happening at a program level, not at a component level. Component version interdependencies need to be tracked and managed.

In the future when there are independent APIs and front-ends there will need to be more complex release strategy, but it's simply overkill right now and the excess complexity is blocking the team's ability to release and test at the velocity required by the business.